edgedb/options.py

```diff
@@ -114,74 +114,158 @@ class State:
+            raise errors.InvalidArgumentError(
+                "with_config() takes from 0 to 1 positional arguments "
+                "but {} were given".format(len(args))
+            )
+        config_dict = args[0] if args else {}
+        config_dict.update(config)
         new_config = self._config.copy()
-        new_config.update(config)
-        return State(
-            module=self._module,
-            aliases=self._aliases,
+        new_config.update(config_dict)
+        return self._new(
+            default_module=self._module,
+            module_aliases=self._aliases,
             config=new_config,
             globals_=self._globals,
         )
 
-    def with_globals(self, **globals_):
+    def resolve(self, name: str) -> str:
+        parts = name.split("::")
+        if len(parts) == 1:
+            return f"{self._module or 'default'}::{name}"
+        elif len(parts) == 2:
+            mod, name = parts
+            mod = self._aliases.get(mod, mod)
+            return f"{mod}::{name}"
+        else:
+            raise errors.InvalidArgumentError(f"Illegal name: {name}")
+
+    def with_globals(self, *args, **globals_):
+        if len(args) > 1:
+            raise errors.InvalidArgumentError(
+                "with_globals() takes from 0 to 1 positional arguments "
+                "but {} were given".format(len(args))
+            )
         new_globals = self._globals.copy()
-        new_globals.update(globals_)
-        return State(
-            module=self._module,
-            aliases=self._aliases,
+        for k, v in (*(args[0] if args else {}).items(), *globals_.items()):
+            new_globals[self.resolve(k)] = v
+        return self._new(
+            default_module=self._module,
+            module_aliases=self._aliases,
+            config=self._config,
+            globals_=new_globals,
+        )
+
+    def without_module_aliases(
+        self, aliases: typing.Optional[typing.Iterable[str]] = None
+    ):
+        if aliases is None:
+            new_aliases = {}
+        else:
+            new_aliases = self._aliases.copy()
+            for alias in aliases:
+                new_aliases.pop(alias, None)
+        return self._new(
+            default_module=self._module,
+            module_aliases=new_aliases,
+            config=self._config,
+            globals_=self._globals,
+        )
+
+    def without_config(
+        self, config_names: typing.Optional[typing.Iterable[str]] = None
+    ):
+        if config_names is None:
+            new_config = {}
+        else:
+            new_config = self._config.copy()
+            for name in config_names:
+                new_config.pop(name, None)
+        return self._new(
+            default_module=self._module,
+            module_aliases=self._aliases,
+            config=new_config,
+            globals_=self._globals,
+        )
+
+    def without_globals(
+        self, global_names: typing.Optional[typing.Iterable[str]] = None
+    ):
```

Comment:
Sure, or just support a signle string?

Reply:
I think it might be too tempting to use `.without_globals('x','y')` and then `.without_globals('x').without_globals('y')` because former crashes (i.e. never learning that it expects an iterator). But it's not a strong opinion.

Reply:
Oh I changed the signature to take `*args` if that makes more sense